### PR TITLE
Migrate `dolt tag` to use SQL

### PR DESCRIPTION
**Pull request by PavelSafronov**
This change updates `dolt tag` command to use only SQL.

Related: #3922

### CODE REVIEW

File: integration-tests/bats/sql-local-remote.bats [hunk 1]
```diff
@@ -736,3 +741,168 @@ SQL
...
+
+  cd ../defaultDB
+  start_sql_server defaultDB
+
+  # get commit hashes
+  headCommit=$(get_commit_hash_at HEAD)
+  secondCommit=$(get_commit_hash_at HEAD~1)
+
+  # show tags
+  run dolt tag
```
Comment (macneale4) [inline integration-tests/bats/sql-local-remote.bats:835]: add the --verbose-engine-setup flag, and check the output to confirm we're really going remote. Alter an invocation in the local context as well.

File: go/cmd/dolt/commands/tag.go [hunk 2]
```diff
@@ -76,6 +86,14 @@ func (cmd TagCmd) Exec(ctx context.Context, commandStr string, args []string, dE
...
+	if err != nil {
+		return handleStatusVErr(err)
+	}
+	if closeFunc != nil {
+		defer closeFunc()
+	}
+
 	// list tags
 	if len(apr.Args) == 0 {
 		var verr errhand.VerboseError
```
Comment (macneale4) [inline go/cmd/dolt/commands/tag.go:99]: There is no reason to have this line three times - lines 99, 115, 133. Just define once at the top of the function

File: integration-tests/bats/sql-local-remote.bats
Comment (macneale4) [inline integration-tests/bats/sql-local-remote.bats:1]: Looks like mosts tests in `commit_tags.bats` could be enabled in local-remote.bash. I see one test that we should skip until checkout is ready.

File: go/cmd/dolt/commands/tag.go [hunk 3]
```diff
@@ -112,81 +136,113 @@ func (cmd TagCmd) Exec(ctx context.Context, commandStr string, args []string, dE
...
+			query = "call dolt_tag(?, ?)"
+			params = []interface{}{tagName, startPoint}
+		} else {
+			query = "call dolt_tag(?, ?, '--author', ?)"
+			params = []interface{}{tagName, startPoint, author}
+		}
 	} else {
-		name, email, err = env.GetNameAndEmail(dEnv.Config)
+		if len(author) == 0 {
+			query = "call dolt_tag('-m', ?, ?, ?)"
```
Comment (macneale4) [inline go/cmd/dolt/commands/tag.go:169]: You should always pass in an author. The stored procedure doesn't have the local user information available to it, so it puts in root@dolt (or something like that) as the author.

**Review by macneale4 — CHANGES_REQUESTED** [review]

File: integration-tests/bats/sql-local-remote.bats (hunk 1 above)
Comment (PavelSafronov) [reply integration-tests/bats/sql-local-remote.bats:835]: Good idead. Adding `--verbose-engine-setup` flag for the first few invocations and verifying that we're getting the results we're expecting.

File: go/cmd/dolt/commands/tag.go (hunk 2 above)
Comment (PavelSafronov) [reply go/cmd/dolt/commands/tag.go:99]: True, though it is nice to have self-contained sections of the code and not have to worry about fall-through. For example, lines 97-111 deal with listing tags, and that's all, no need to look elsewhere.
That being said, most of these blocks are just doing some param validation and then delegate to a small function, so I'm going to move the logic into those functions. Should make things a bit cleaner.

File: integration-tests/bats/sql-local-remote.bats
Comment (PavelSafronov) [reply integration-tests/bats/sql-local-remote.bats:1]: Enabling.

File: go/cmd/dolt/commands/tag.go (hunk 3 above)
Comment (PavelSafronov) [reply go/cmd/dolt/commands/tag.go:142]: Hmm, that's not what I'm observing. This BATS test just succeeded for me. I'm specifically checking for local/remote execution, not specifying an author, and verifying that the author is `Bats Tests <bats@email.fake>`:

```sh
  # add tag without message
  run dolt --verbose-engine-setup tag v1
  [ $status -eq 0 ]
  [[ $output =~ "verbose: starting remote mode" ]] || false

  # list tags and check new tag is present
  run dolt tag
  [ $status -eq 0 ]
  [[ $output =~ "v1" ]] || false

  # list tags with verbose flag and check new tag is present
  run dolt --verbose-engine-setup tag --verbose
  [ $status -eq 0 ]
  [[ $output =~ "v1"$'\t'"$headCommit" ]] || false
  [[ $output =~ "Tagger: Bats Tests <bats@email.fake>" ]] || false
  [[ $output =~ "verbose: starting remote mode" ]] || false
```

Comment (macneale4) [reply go/cmd/dolt/commands/tag.go:142]: Huh. Interesting. I'm convinced by the test :)

**Review by macneale4 — APPROVED** [review]
```
    __4___
 _  \ \ \ \
<'\ /_/_/_/
 ((____!___/)
  \0\0\0\0\/
 ~~~~~~~~~~~
```

**Pull request merged**